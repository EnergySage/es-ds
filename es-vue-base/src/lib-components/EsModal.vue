<template>
    <b-modal
        :id="id"
        :size="size"
        :scrollable="scrollable"
        :visible="visible"
        title-class="h2"
        :hide-footer="hideFooter"
        centered
        static
        class="modal-lightbox"
        v-bind="$attrs"
        v-on="$listeners">
        <template #modal-title>
            <slot name="modal-title" />
        </template>
        <template #modal-header-close>
            <XIcon
                class="text-gray-900"
                height="26px"
                width="26px" />
        </template>
        <slot />
        <template #modal-footer>
            <slot name="modal-footer" />
        </template>
    </b-modal>
</template>

<script lang="js">
import { BModal } from 'bootstrap-vue';
import XIcon from '@/src/lib-icons/x.vue';

export default {
    name: 'EsModal',
    components: {
        BModal,
        XIcon,
    },
    props: {
        /**
         * ID
         */
        id: {
            type: String,
            required: true,
        },
        /**
         * Size
         */
        size: {
            type: String,
            default: 'md',
            validator: (val) => ['sm', 'md', 'lg', 'xl'].includes(val),

        },
        /**
         * Scrollable
         */
        scrollable: {
            type: Boolean,
            default: true,
        },
        /**
         * Visible
         */
        visible: {
            type: Boolean,
            default: false,
        },
        /**
         * Hide Footer
         */
        hideFooter: {
            type: Boolean,
            default: false,
        },
    },
};
</script>

<style lang="scss">
@import '~@energysage/es-bs-extends/scss/includes';

// TODO: Move this into es-bs-extends ?
.modal-lightbox {

    @include media-breakpoint-down(md) {
        .modal-dialog {
            margin: 0;
            min-height: 100%;
            min-width: 100%;

            .modal-content {
                border-radius: 0;
            }
        }
    }
}

</style>
