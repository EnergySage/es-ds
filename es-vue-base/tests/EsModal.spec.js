import { mount } from '@vue/test-utils';
import EsModal from '@/src/lib-components/EsModal.vue';
import jestVue from '@/tests/jest.vue.config';

const slots = {
    'modal-title': 'Reviews',
    default: '<h3>Great experience!</h3>',
    'modal-footer': 'Close',
};

describe('EsModal', () => {
    // Basic test; should exist for most components
    test('<EsModal />', async () => {
        const wrapper = mount(EsModal, {
            attachTo: document.body,
            ...jestVue,
            propsData: {
                id: 'es-modal',
            },
        });
        await wrapper.vm.$nextTick();
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    // Test slots are there
    test('<EsModal><title><body><footer></EsModal>', async () => {
        const wrapper = mount(EsModal, {
            attachTo: document.body,
            ...jestVue,
            propsData: {
                id: 'es-modal',
            },
            slots,
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.html()).toContain('Reviews');
        expect(wrapper.html()).toContain('<h3>Great experience!</h3>');
        expect(wrapper.html()).toContain('Close');
    });

    // Test custom props/default prop values are there
    test('<EsModal hideFooter />', async () => {
        const wrapper = mount(EsModal, {
            attachTo: document.body,
            ...jestVue,
            propsData: {
                id: 'es-modal',
            },
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.props('hideFooter')).toBe(false);
        expect(wrapper.props('scrollable')).toBe(true);
    });
});
