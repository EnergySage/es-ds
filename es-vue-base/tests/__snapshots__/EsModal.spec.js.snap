// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsModal <EsModal /> 1`] = `
"<div id="es-modal___BV_modal_outer_" style="position: absolute; z-index: 1040;" class="modal-lightbox">
  <transition-stub enterclass="" leaveclass="" entertoclass="" leavetoclass="" enteractiveclass="" leaveactiveclass="">
    <div id="es-modal" role="dialog" aria-hidden="true" aria-labelledby="es-modal___BV_modal_title_" aria-describedby="es-modal___BV_modal_body_" class="modal fade" style="display: none;">
      <div class="modal-dialog modal-md modal-dialog-centered modal-dialog-scrollable">
        <!---->
        <div id="es-modal___BV_modal_content_" tabindex="-1" class="modal-content">
          <header id="es-modal___BV_modal_header_" class="modal-header">
            <h5 id="es-modal___BV_modal_title_" class="modal-title h2"></h5><button type="button" aria-label="Close" class="close"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" style="height: 26px; width: 26px;" class="text-gray-900">
                <path fill-rule="evenodd" clip-rule="evenodd" d="M5.293 5.293a1 1 0 0 1 1.414 0L12 10.586l5.293-5.293a1 1 0 1 1 1.414 1.414L13.414 12l5.293 5.293a1 1 0 0 1-1.414 1.414L12 13.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L10.586 12 5.293 6.707a1 1 0 0 1 0-1.414Z"></path>
              </svg></button>
          </header>
          <div id="es-modal___BV_modal_body_" class="modal-body"> </div>
          <footer id="es-modal___BV_modal_footer_" class="modal-footer"><button type="button" class="btn btn-secondary">Cancel</button><button type="button" class="btn btn-primary">OK</button></footer>
        </div>
        <!---->
      </div>
    </div>
  </transition-stub>
  <transition-stub name="" css="true" enterclass="" leaveclass="show" entertoclass="show" leavetoclass="" enteractiveclass="fade" leaveactiveclass="fade">
    <!---->
  </transition-stub>
</div>"
`;

exports[`EsModal <EsModal hideFooter /> 1`] = `
"<div id="es-modal___BV_modal_outer_" style="position: absolute; z-index: 1040;" class="modal-lightbox">
  <transition-stub enterclass="" leaveclass="" entertoclass="" leavetoclass="" enteractiveclass="" leaveactiveclass="">
    <div id="es-modal" role="dialog" aria-hidden="true" aria-labelledby="es-modal___BV_modal_title_" aria-describedby="es-modal___BV_modal_body_" class="modal fade" style="display: none;">
      <div class="modal-dialog modal-md modal-dialog-centered modal-dialog-scrollable">
        <!---->
        <div id="es-modal___BV_modal_content_" tabindex="-1" class="modal-content">
          <header id="es-modal___BV_modal_header_" class="modal-header">
            <h5 id="es-modal___BV_modal_title_" class="modal-title h2"></h5><button type="button" aria-label="Close" class="close"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" style="height: 26px; width: 26px;" class="text-gray-900">
                <path fill-rule="evenodd" clip-rule="evenodd" d="M5.293 5.293a1 1 0 0 1 1.414 0L12 10.586l5.293-5.293a1 1 0 1 1 1.414 1.414L13.414 12l5.293 5.293a1 1 0 0 1-1.414 1.414L12 13.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L10.586 12 5.293 6.707a1 1 0 0 1 0-1.414Z"></path>
              </svg></button>
          </header>
          <div id="es-modal___BV_modal_body_" class="modal-body"> </div>
          <footer id="es-modal___BV_modal_footer_" class="modal-footer"><button type="button" class="btn btn-secondary">Cancel</button><button type="button" class="btn btn-primary">OK</button></footer>
        </div>
        <!---->
      </div>
    </div>
  </transition-stub>
  <transition-stub name="" css="true" enterclass="" leaveclass="show" entertoclass="show" leavetoclass="" enteractiveclass="fade" leaveactiveclass="fade">
    <!---->
  </transition-stub>
</div>"
`;

exports[`EsModal <EsModal><title><body><footer></EsModal> 1`] = `
"<div id="es-modal___BV_modal_outer_" style="position: absolute; z-index: 1040;" class="modal-lightbox">
  <transition-stub enterclass="" leaveclass="" entertoclass="" leavetoclass="" enteractiveclass="" leaveactiveclass="">
    <div id="es-modal" role="dialog" aria-hidden="true" aria-labelledby="es-modal___BV_modal_title_" aria-describedby="es-modal___BV_modal_body_" class="modal fade" style="display: none;">
      <div class="modal-dialog modal-md modal-dialog-centered modal-dialog-scrollable">
        <!---->
        <div id="es-modal___BV_modal_content_" tabindex="-1" class="modal-content">
          <header id="es-modal___BV_modal_header_" class="modal-header">
            <h5 id="es-modal___BV_modal_title_" class="modal-title h2">Reviews</h5><button type="button" aria-label="Close" class="close"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" style="height: 26px; width: 26px;" class="text-gray-900">
                <path fill-rule="evenodd" clip-rule="evenodd" d="M5.293 5.293a1 1 0 0 1 1.414 0L12 10.586l5.293-5.293a1 1 0 1 1 1.414 1.414L13.414 12l5.293 5.293a1 1 0 0 1-1.414 1.414L12 13.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L10.586 12 5.293 6.707a1 1 0 0 1 0-1.414Z"></path>
              </svg></button>
          </header>
          <div id="es-modal___BV_modal_body_" class="modal-body">
            <h3>Great experience!</h3>
          </div>
          <footer id="es-modal___BV_modal_footer_" class="modal-footer">Close</footer>
        </div>
        <!---->
      </div>
    </div>
  </transition-stub>
  <transition-stub name="" css="true" enterclass="" leaveclass="show" entertoclass="show" leavetoclass="" enteractiveclass="fade" leaveactiveclass="fade">
    <!---->
  </transition-stub>
</div>"
`;
